### PR TITLE
17 canvas

### PR DESCRIPTION
**Pull request by magdalena-u**
chapter 17: Drawing on Canvas:
17_1 Shapes
17_2 A pie chart
17_3 A bouncing ball

### CODE REVIEW

File: 17_1_shape.js [hunk 1]
```diff
@@ -0,0 +1,64 @@
+const ctx = document.querySelector('canvas').getContext('2d');
```
Comment (mateusz-bielak) [inline 17_1_shape.js]: Maybe, instead of `ctx` name it just `canvas`? :)

File: 17_1_shape.js [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+// A trapezoid
+
+ctx.strokeStyle = 'black';
```
Comment (mateusz-bielak) [inline 17_1_shape.js]: Maybe collect all these expressions into functions like `generateTrapezoid` etc.

File: 17_1_shape.js [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+//A zigzagging line
+
+ctx.strokeStyle = 'black';
+ctx.moveTo(50, 50);
+let counter = 50;
+
+for (let i = 0; i <= 12; i++) {
+    counter += 25;
+    if (i % 2 === 0) {
```
Comment (mateusz-bielak) [inline 17_1_shape.js]: You can easily use a ternary operator here.

File: 17_1_shape.js [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+const x = 250;
+const y = 250;
+
+ctx.strokeStyle = 'black';
+ctx.beginPath();
+ctx.moveTo(x, y);
+
+for (let i = 0; i < 200; i++) {
+    const angle = (i * Math.PI) / 20;
+    let radius = 30 + i;
```
Comment (mateusz-bielak) [inline 17_1_shape.js]: Why `let`?

File: 17_1_shape.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    ctx.lineTo(x + Math.cos(angle) * radius, y + Math.sin(angle) * radius);
```
Comment (mateusz-bielak) [inline 17_1_shape.js]: What are you calculating here? Maybe put that into a variable like `myPerfectRadius` ;)

File: 17_1_shape.js (hunk 1 above)
Comment (mateusz-bielak) [reply 17_1_shape.js]: Oh, you are using `getContext` also. So maybe `context`? Using shortcuts isn't a good pattern usually.

File: 17_2_chart.js [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+];
+
+const ctx = document.querySelector('canvas').getContext('2d');
+const total = results.reduce((sum, { count }) => sum + count, 0);
+let currentAngle = -0.5 * Math.PI;
+const centerX = 300;
+const centerY = 150;
+
+// Add code to draw the slice labels in this loop.
+for (let result of results) {
```
Comment (mateusz-bielak) [inline 17_2_chart.js]: Please, try to group expressions in that loop. Also, maybe you will be able to extract some parts into separate functions then.

File: 17_2_chart.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    const sliceAngle = (result.count / total) * 2 * Math.PI;
+    ctx.beginPath();
+    ctx.arc(centerX, centerY, 100, currentAngle, currentAngle + sliceAngle);
+    currentAngle += sliceAngle;
+    const middleAngle = currentAngle + 0.5 * sliceAngle;
+    const textX = Math.cos(middleAngle) * 120 + centerX;
+    const textY = Math.sin(middleAngle) * 120 + centerY;
+    if (middleAngle < 0) {
```
Comment (mateusz-bielak) [inline 17_2_chart.js]: Use ternary, please ❤️

File: 17_2_chart.js [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+    const sliceAngle = (result.count / total) * 2 * Math.PI;
+
+    canvas.beginPath();
+    canvas.arc(centerX, centerY, 100, currentAngle, currentAngle + sliceAngle);
+    currentAngle += sliceAngle;
+
+    const middleAngle = currentAngle + 0.5 * sliceAngle;
+    const textX = Math.cos(middleAngle) * 120 + centerX;
+    const textY = Math.sin(middleAngle) * 120 + centerY;
+
```
Comment (michal-karbasz) [inline 17_2_chart.js:19]: this might need some adjustments

**Review by michal-karbasz — APPROVED** [review]

File: 17_1_shape.js (hunk 1 above)
Comment (michal-karbasz) [reply 17_1_shape.js]: it's a pattern that canvas users will recognize, though

File: 17_1_shape.js (hunk 2 above)
Comment (michal-karbasz) [reply 17_1_shape.js]: yup, if you'll try to open that file you'll see that variable declarations for `x` are duplicated - but them into separate scopes

**Pull request merged**